packages/x-date-pickers/src/internals/hooks/useField/useFieldCharacterEditing.ts

```diff
@@ -0,0 +1,442 @@
+        return queryResponse;
+      }
+    }
+
+    const queryResponse = getFirstSectionValueMatchingWithQuery(cleanKeyPressed, activeSection);
+    if (isQueryResponseWithoutValue(queryResponse) && !queryResponse.saveQuery) {
+      setQuery(null);
+    } else {
+      setQuery({
+        sectionIndex,
+        value: cleanKeyPressed,
+        dateSectionName: activeSection.dateSectionName,
+      });
+    }
+
+    if (isQueryResponseWithoutValue(queryResponse)) {
+      return null;
+    }
+    return queryResponse;
+  };
+
+  const applyLetterEditing: CharacterEditingApplier<TDate, TSection> = (params) => {
+    const getFirstSectionValueMatchingWithQuery: QueryApplier<TSection> = (
+      queryValue,
+      activeSection,
+    ) => {
+      switch (activeSection.dateSectionName) {
+        case 'month': {
+          const getMonthResponse = (format: string): ReturnType<QueryApplier<TSection>> => {
+            const matchingMonths = utils
+              .getMonthArray(utils.date()!)
+              .map((month) => utils.formatByString(month, format!))
+              .filter((month) => month.toLowerCase().startsWith(queryValue));
+
+            if (matchingMonths.length === 0) {
+              return { saveQuery: false };
+            }
+
+            return {
+              sectionValue: matchingMonths[0],
+              shouldGoToNextSection: matchingMonths.length === 1,
+            };
+          };
+
+          if (activeSection.contentType === 'letter') {
+            return getMonthResponse(activeSection.formatValue);
+          }
+
+          // When editing a digit-format month and the user presses a letter,
+          // We can support the letter editing by using the letter-format month and re-formatting the result.
+          // We just have to make sure that the default month format is a letter format,
+          if (
+            getDateSectionConfigFromFormatToken(utils, utils.formats.month).contentType === 'letter'
+          ) {
+            const monthResponse = getMonthResponse(utils.formats.month);
+            if (isQueryResponseWithoutValue(monthResponse)) {
+              return { saveQuery: false };
+            }
+
+            const formattedValue = changeSectionValueFormat(
+              utils,
+              monthResponse.sectionValue,
+              utils.formats.month,
+              activeSection.formatValue,
+            );
+
+            return {
+              ...monthResponse,
+              sectionValue: formattedValue,
+            };
+          }
+
+          return { saveQuery: false };
+        }
+
+        case 'meridiem': {
+          const now = utils.date()!;
+          const sectionValue = [utils.endOfDay(now), utils.startOfDay(now)]
+            .map((date) => utils.formatByString(date, activeSection.formatValue))
+            .find((meridiem) => meridiem.toLowerCase().startsWith(queryValue));
+
+          if (sectionValue == null) {
+            return { saveQuery: false };
+          }
```

Comment:
Could be buggy if there is a local with a meridian starting with the same letter

But not sure it worth to handle such a case

Reply:
This line is only used when no meridiem matches, so we should have no issue even with multiple meridiems starting with the same letter